Zotero/Controllers/OpenItemsController.swift

```diff
@@ -0,0 +1,235 @@
+        guard items != self.items else { return }
+        self.items = items
+        observable.on(.next(items))
+    }
+    
+    func open(_ item: Item) {
+        DDLogInfo("OpenItemsController: opened item \(item)")
+        guard items.last != item else { return }
+        // Since item not last, will certainly need to be appended in the end
+        if let index = items.firstIndex(of: item) {
+            // Item already open, remove it to be appended
+            items.remove(at: index)
+            DDLogInfo("OpenItemsController: moving already opened item \(item) to most recent")
+        } else {
+            DDLogInfo("OpenItemsController: appending newly opened item \(item) as most recent")
+        }
+        items.append(item)
+        observable.on(.next(items))
+    }
+    
+    @discardableResult
+    func restore(_ item: Item, using presenter: OpenItemsPresenter) -> Bool {
+        guard let (library, url) = load(item: item) else { return false }
+        presentItem(item, at: url, library: library, using: presenter)
+        return true
+    }
+    
+    @discardableResult
+    func restoreMostRecentlyOpenedItem(using presenter: OpenItemsPresenter) -> Item? {
+        // Will restore most recent opened item still present, or none if all fail
+        DDLogInfo("OpenItemsController: restoring most recently opened item using presenter \(presenter)")
+        var itemsChanged: Bool = false
+        defer {
+            if itemsChanged {
+                observable.on(.next(items))
+            }
+        }
+        var item: Item?
+        var library: Library?
+        var url: URL?
+        while let _item = items.last {
+            if let (_library, _url) = load(item: _item) {
+                item = _item
+                library = _library
+                url = _url
+                break
+            }
+            DDLogWarn("OpenItemsController: removing not loaded item \(_item)")
+            _ = items.removeLast()
+            itemsChanged = true
+        }
+        guard let item, let library, let url else { return nil }
+        presentItem(item, at: url, library: library, using: presenter)
+        return item
+    }
+    
+    func deferredOpenItemsMenuElement(disableOpenItem: Bool, itemActionCallback: @escaping (Item, UIAction) -> Void) -> UIDeferredMenuElement {
+        UIDeferredMenuElement { [weak self] elementProvider in
+            guard let self else {
+                elementProvider([])
+                return
+            }
+            var actions: [UIAction] = []
+            let items = self.items
+            let openItem: Item? = disableOpenItem ? items.last : nil
+            var itemTuples: [(Item, RItem)] = []
+            do {
+                try dbStorage.perform(on: .main) { coordinator in
+                    for item in items {
+                        do {
+                            let rItem = try coordinator.perform(request: ReadItemGloballyDbRequest(key: item.key))
```

Comment:
The `Item` enum holds `LibraryIdentifier` too, so why is there a need for `ReadItemGloballyDbRequest`?

Reply:
Fixed in https://github.com/zotero/zotero-ios/pull/782/commits/114f136cd6bffe050c751ba132ae2a30d7511e0c to use proper request. In an interim version, based on my mistake of `key` being unique in the database, I had a single request that read all items at once, which I later modified to a loop, so that's how I got there 😅